core/keyboard_nav/key_map.js

```diff
@@ -29,68 +29,62 @@ goog.provide('Blockly.user.keyMap');
 
 /**
  * Holds the serialized key to key action mapping.
- * @type {Object<string, Blockly.KeyAction>}
+ * @type {Object<string, Blockly.Action>}
  */
 Blockly.user.keyMap.map_ = {};
 
 /**
- * Holds the list of actions that can be executed using the keyboard.
- * @type {Array<Blockly.KeyAction>}
+ * List of modifier keys checked when serializing the key event.
+ * @type {Array<string>}
  */
-Blockly.user.keyMap.actions_ = [];
+Blockly.user.keyMap.modifierKeys = ['Shift','Control','Alt','Meta'];
 
 /**
- * Update the key map and actions array to contain the new action.
+ * Update the key map to contain the new action.
  * @param {!string} keyCode The key code serialized by the serializeKeyEvent.
- * @param {!Blockly.KeyAction} action The action to be executed when the keys
+ * @param {!Blockly.Action} action The action to be executed when the keys
  *     corresponding to the serialized key code is pressed.
  * @package
  */
 Blockly.user.keyMap.setActionForKey = function(keyCode, action) {
-  var key = Blockly.user.keyMap.getKeyByActionName(action.name);
+  var oldKey = Blockly.user.keyMap.getKeyByAction(action);
   // If the action already exists in the key map remove it and add the new mapping.
-  if (key) {
-    delete Blockly.user.keyMap.map_[key];
+  if (oldKey) {
+    delete Blockly.user.keyMap.map_[oldKey];
   }
   Blockly.user.keyMap.map_[keyCode] = action;
-  // If the action is not in the actions array add it.
-  if (Blockly.user.keyMap.actions_.indexOf(action) === -1) {
-    Blockly.user.keyMap.actions_.push(action);
-  }
 };
 
 /**
- * Creates a new key map and a new actions array.
- * @param {Object<string, Blockly.KeyAction>} keyMap The object holding the key
+ * Creates a new key map.
+ * @param {Object<string, Blockly.Action>} keyMap The object holding the key
  *     to action mapping.
  * @package
  */
 Blockly.user.keyMap.setKeyMap = function(keyMap) {
-  var keys = Object.keys(keyMap);
-  // Update the actions array to include any new actions found in the new key map.
-  for (var i = 0; i < keys.length; i++) {
-    var action = keyMap[keys[i]];
-    if (Blockly.user.keyMap.actions_.indexOf(action) === -1) {
-      Blockly.user.keyMap.actions_.push(action);
-    }
-  }
   Blockly.user.keyMap.map_ = keyMap;
 };
 
 /**
  * Gets the current key map.
- * @return {Object<string,Blockly.KeyAction>} The object holding the key to
+ * @return {Object<string,Blockly.Action>} The object holding the key to
  *     action mapping.
  * @package
  */
 Blockly.user.keyMap.getKeyMap = function() {
-  return Blockly.user.keyMap.map_;
+  var mapClone = {};
+  var keys = Object.keys(Blockly.user.keyMap.map_);
+  for (var i = 0, key; key = keys[i]; i++) {
+    var action = Blockly.user.keyMap.map_[key];
+    mapClone[key] = new Blockly.Action(action.name, action.desc, action.func);
```

Comment:
It's probably sufficient to return the original action in a copy of the map. If they want to modify the action directly they have access to it on the namespace, so it's less likely they'll get it here for that purpose.